src/Juvix/Compiler/Core/Translation/FromInternal.hs

```diff
@@ -437,28 +490,31 @@ goExpression' = \case
   Internal.ExpressionApplication a -> goApplication a
   Internal.ExpressionSimpleLambda l -> goSimpleLambda l
   Internal.ExpressionLambda l -> goLambda l
-  Internal.ExpressionFunction f -> unsupported ("goExpression function: " <> show (Loc.getLoc f))
+  e@(Internal.ExpressionFunction {}) -> goFunction (Internal.unfoldFunType e)
   Internal.ExpressionHole h -> error ("goExpression hole: " <> show (Loc.getLoc h))
-  Internal.ExpressionUniverse u -> error ("goExpression universe: " <> show (Loc.getLoc u))
+  Internal.ExpressionUniverse {} -> return (mkUniv' (fromIntegral smallLevel))
+
+goFunction ::
+  forall r.
+  Members '[InfoTableBuilder, Reader InternalTyped.TypesTable, Reader Internal.InfoTable, Reader IndexTable] r =>
+  ([Internal.FunctionParameter], Internal.Expression) ->
+  Sem r Node
+goFunction (params, returnTypeExpr) = do
+  foldr f (goExpression returnTypeExpr) params
+  where
+    f :: Internal.FunctionParameter -> Sem r Node -> Sem r Node
+    f param acc = do
+      paramBinder <- Binder (param ^. Internal.paramName) <$> goExpression (param ^. Internal.paramType)
+      case param ^. Internal.paramName of
+        Nothing -> mkPi mempty paramBinder <$> acc
```

Comment:
we have the shortcut `mkPi'` when the info is empty

Reply:
`mkPi'` doesn't take a binder so can't be used at this call site.

I thought about adding a new variant of `mkPi` that takes a binder but I decided against it. The pattern we have is that `'` variants allow you to omit the `Info` argument, which is easy to remember. If we have variants for "no Info, Type" and "no Info, Binder" I think I'd always have to look at the definition to figure out which to use.

I suppose an alternative is to make `mkPi'` take a binder and fix the call sites of the existing calls, or to add a new variant called `mkPiBinder'`. What do you think?

Reply:
You are right!

> I suppose an alternative is to make mkPi' take a binder and fix the call sites of the existing calls, or to add a new variant called mkPiBinder'. What do you think?
I think it makes more sense to modify mkPi' so that it takes a binder, but it is not a strong preference so I'll leave it up to you

Reply:
I tried updating `mkPi'` to take a binder but that made the call sites of the existing binderless `mkPi'` calls quite noisy. On balance I think it's best to use the full `mkPi` at the call site instead of changing the others. This could change if we start using the Binder form more.